[F3D] Add material bleeding to FMesh GfxLists

jesusyoshi54 opened the pull request:
This commit will allow users to export mesh gfx lists with material bleeding. What this will do is remove unnecessary DL cmds inside the gfxlist, and make all the DL cmds part of one gfx list. This option is added as a checkbox in the global f3d settings.

This commit also adjusts texture scrolls for sm64 so that they work with mat bleeding, and fixes a bug where not setting a texture still exported a tile scroll for that material (which would cause UB in game).

This commit has been tested to successfully export multiple test sm64 levels and has been verified to work on console. No testing has been done for OOT.

YoshiMilkman commented:
Currently getting compilation errors when trying to export my object geolayout with the material bleed checkbox turned on. The export works perfectly fine with it off.
![Capture](https://user-images.githubusercontent.com/77122601/204342288-e349da72-2e34-407b-905e-3fba5d654c64.PNG)

jesusyoshi54 commented:
Ok I understand, there is currently an issue with the geo layout writer not knowing which DLs to use when multiple layers are written. I will fix this in an upcoming commit.

jesusyoshi54 commented:
Naming issue has been resolved. Geo layouts should now properly reference the correct bled display list.
Please test with the latest commit.

YoshiMilkman commented:
It compiles now, thank you.

kurethedead commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -1787,7 +1790,9 @@ def get_tex_sts_code(tex: FSetTileSizeScrollField, tex_num: int, cmd_num: int):
 ):
-    func = f"void scroll_sts_{mat}()"
+    # remove any offsets from mat name
+    func_name = mat.split("+")[0].strip()
+    func = f"void scroll_sts_{func_name}()"
```
This is a bit hacky, also related to tile scrolling bleeds if I'm understanding correctly.

kurethedead reviewed (changes requested):
I tried testing a map export, which fails to compile. There are a couple other errors as well - I think this should have been tested more before being PR'd. However, the main issue I have is organization. I've mentioned this before, but this PR adds a bunch of `to_c_inline` functions which does display list building when it should only be doing c exporting, adding a lot of unnecessary code in the process. In `FMesh.bleed()`, instead of saving data to a `bleed_gfx` class, the final inlined DL can just be built there and saved over `FMesh.draw`. `FMesh` can then use `FMesh.inline` to determine which data to export in `FMesh.to_c()`. This removes the need for `to_c_inline` functions, and also the various other inline checks that are done when handling FMesh writing. 

The tile scrolling bleed is a bit hacky, although thats more due to the original tile scrolling being inflexible. I have some issue with how that's organized as well, although that's a whole other thing. Instead, bleed() could save an OrderedDict[FMaterial, tileCommandIndices] to `FMesh`. `FMesh.to_c_vertex_scroll` can then be refactored so that it executes code that processes this dict along side the fully inlined DL if inline is true. This would remove the need to create new `FMaterials`.

I've made a couple comments that might not be relevant anymore if `to_c_inline` is removed. A couple of the errors here are also irrelevant in this case.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2711,6 +2752,133 @@ def save_binary(self, romfile, f3d, segments):
+            if triGroup.bleed.bleed_tile_scroll:
+                scr_mat = FMaterial("scr", DLFormat.Static)
+                for GfxList in triGroup.bleed.bled_tex:
+                    scr_mat.material.commands.extend(GfxList.commands)
+                scr_mat.material.name = f"{self.name} + {size // 8}"
+                scr_mat.name = f"{self.name} + {size // 8}"
+                scr_mat.scrollData = triGroup.fMaterial.scrollData
+                scr_mat.texturesLoaded = triGroup.fMaterial.texturesLoaded
```
If the tile scroll code is refactored I could avoid this, but I believe that would be more messy, or would require implementing entirely new tile scroll code separate for inline materials.
I can make an attempt at it if that's the preferred direction but that was also the initial way I tried when making this PR, and it was much more complicated.

kurethedead replied:
This current solution is already pretty messy itself. The original tile scrolling code is pretty inflexible, but unless you want to refactor that too then its better to handle inline tiles scrolls as a separate case, which it is.

jesusyoshi54 replied:
I believe the best way to change it would still be to make edits to TriGroup data, though instead of making a new material I can edit the scroll generation to make use of inline attributes so that the correct data is generated.
Here is the list of items Id need to change in tile scroll generation to work with inline scrolls.
1. The function naming convention will need to be remade to come from the TriGroup/FMesh.
2. The cmds parsed to find the scroll index will have to come from TriGroup.bleed or some other TriGroup attribute.
3. The offsets used in the tile scroll will have to be relative to the start of the TriGroups parsed cmds, TriGroup.tile_offset perhaps.

Overall I think it would be similar to the current implementation except I would be moving from editing the material in the TriGroup to editing the TriGroup itself and the scroll generation

kurethedead replied:
Why not calculate the offsets in `bleed()` as you're building the display list? This is what I was referring to when talking about having a OrderedDict[FMaterial, tileCommandIndices] as an FMesh property, which would store indices into the inlined DL. This way you don't need any extra TriGroup properties at all.

jesusyoshi54 replied:
There is no reason it can't be done I just feel like it makes more sense to have it be defined in the TriGroup because that is the object that holds the material and generates the individual scroll.
It would also make the new inline code only require a few differences from the original code  which I believe will make it easier to maintain in the future.

kurethedead replied:
I'm not sure I understand - once you inline the DL, you're storing indices into `FMesh.draw`. These indices don't refer to any particular FTriGroup.

jesusyoshi54 replied:
I would be storing the indices in the bleed_gfx object, which is a member of the triGroup. As far as material bleeding goes, it is based on triGroups, and the material is unique to that triGroup.
It is correct that the new display list is a part of the FMesh object now, but the owner of the bled gfx is the triGroup, which is where the tile indices will go. To me it makes more sense to synchronizes one scroll to each material (which is what already happens) than to make entirely new behavior that places all scrolls from the given indices in a FMesh.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2711,6 +2752,133 @@ def save_binary(self, romfile, f3d, segments):
+            triGroup.bleed = bleed_gfx(bleed_mat, bleed_tex, bleed_scr)
+        return LastMat
+
+    def bleed_textures(self, mat, LastMat):
```
A lot of type hints are not added due to file ordering.
Maybe there is another way, but writing a type hint of FMaterial results in an undefined symbol error.

kurethedead replied:
You can add quotes around your type to forward reference it.

kurethedead commented:
I'm not sure, vtx culling was as far as I got. I also forgot to mention, I have a tile scrolling issue: I'm trying to export a mesh with two materials, one with s/t tile scrolling and another with s tile scrolling, but only one of the material's tile scrolling code is being exported.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2711,6 +2752,133 @@ def save_binary(self, romfile, f3d, segments):
+            LastList = LastMat.material.commands
+            cnt = 0
+            for j, cmd in enumerate(GfxList.commands):
+                if cmd.bleed(LastList, commands_bled.commands, j - cnt):
+                    if cmd in LastList:
+                        commands_bled.commands.pop(j - cnt)  # list gets smaller as I pop, so modify index by num popped
+                        cnt += 1
```
I did this because I was worried about duplicate cmds within the list being accidentally removed. An example is a sync, which can be present in the same GfxList multiple times. I may want to remove just one of them, but the second one only: a case where I would have to remove by `list.pop(index)`.
This convention would also result in me having to keep track of the amount of removed indices to properly reference the correct one.

kurethedead replied:
Maybe you can just replace the command with a `None`, then filter out all the `None`s in the array at the end? That way the command index lines up between all the display lists.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2229,11 +2250,22 @@ def addMesh(self, name, namePrefix, drawLayer, isSkinned, contextObj):
+        return
+
     def endDraw(self, fMesh, contextObj):
+        if not fMesh.inline:
+            fMesh.draw.commands.append(SPEndDisplayList())
         self.onEndDraw(fMesh, contextObj)
-        fMesh.draw.commands.append(SPEndDisplayList())
 
     def onEndDraw(self, fMesh, contextObj):
+        if fMesh.inline:
+            fMesh.bleed()
         return
```
`FModel.onEndDraw` is used in the codebase by itself (at least once), it doesn't do anything but it isn't entirely pointless.

kurethedead commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -5176,24 +5419,51 @@ class DPFullSync(GbiMacro):
 
+    # will be bled if there are two of these syncs, at most only one pipesync
+    # is ever required after rendering triangles, and subsequent cmds rdp attr changes
+    def bleed(self, LastGfxList: GfxList, CurList: GfxList, curIndex: int):
```
For these sync commands, it seems like these bleed functions always return True?

jesusyoshi54 replied:
Ok this is a silly bug lol. It should be `return False` on the outside if the loop.

jesusyoshi54 commented:
I have fixed the issues with culling exporting and tile scrolling and improved the names for a lot of variables.
I have also fixed some inconsistencies with cmds that were ignored when added to fMesh.draw while creating materials for large textures.
With these changes I was also able to clean up some no longer useful code.

Now exporting is done by making use of the fMesh.draw object, and it is appended to using the triGroup.bleed data after bleeding for that triGroup is finished. I also use this time to add in the tile scroll indices (or at least where to start searching for them). The theory of operation is still mostly the same, it is just organized differently to take the work out of the `to_c_inline` method and put it into the `bleed` method.

I have also written a new tile scrolling method for inline materials. This method is basically the same as the original, but it uses data from the triGroup to select the correct GfxList, and to find the tile indices.

I was not able to completely get rid of the `to_c_inline` method as triGroup writing is still out of the order needed for inline material drawing.

As far as the decision to keep `triGroup.bleed` when it could probably be deprecated by this point, it is mostly me planning for the future with additional modding support. I believe keeping the bled gfx available for further post processing or mutation may end up being very convenient for certain types of exports, and for additional searching for dynamic DLs beyond sm64 tile scrolls.

kurethedead reviewed (changes requested):
To me it seems like the `to_c_inline` can be removed, and instead `FMesh.inline` can be used inside `to_c` to get the desired result. I'm still not seeing the point in saving bled data intermediately in FTriGroup, as everything can still be done in `bleed()`. If you want to add additional functionality that uses that info, then you can add callbacks inside `bleed()` that are implemented by game specific classes which pass in the desired display lists.

jesusyoshi54 commented:
I can add an onBleedEnd method and store the bleed prop but I use the bleed to hold the tile scroll information, so it needs to be stored somewhere.
It is much easier to hold the information in the triGroup over the FMesh.draw list because I can parse the individual tile for the cmds I need, and I am not required to commit to identifying only tile scroll indices in cases where other types of scrolls are made in the future

kurethedead commented on fast64_internal/f3d/f3d_writer.py:
```diff
@@ -554,7 +555,16 @@ def saveStaticModel(
     for drawLayer, fMesh in fMeshes.items():
-        if revertMatAtEnd:
+        #inline requires the mesh to revert a few cmds for safety
+        if fMesh.inline:
+            # basic revert adds one too many syncs for inline
+            revertMatAndEndDraw(tmpList := GfxList("", GfxListTag.Draw, DLFormat.Static), [DPSetRenderMode(fModel.getRenderMode(drawLayer), None), DPSetCycleType("G_CYC_1CYCLE"), DPSetTextureLUT("G_TT_NONE")])
+            tmpList.commands.remove(DPPipeSync())
+            
+            fModel.endDraw(fMesh, obj)
+            # add revert list without sync
+            fMesh.draw.commands.extend(tmpList.commands)
```
Honestly, I wouldn't even use that function at all in this case, since its small enough and you're modifying its output anyway. Or it can be moved into f3d_gbi.

kurethedead commented:
I was thinking of a callback that would be added inside the loop iterating over triGroups, not at the end. I don't see why the bleed data would be needed after bleeding is done, since the final output of bleed is to merge all the bled display lists together. Also, editing that inlined DL would invalidate tile indices. I still disagree with the location of tile scrolling indices (its a property of a specific display list first before being a property of a material), but you can just store that in the triGroup itself. Also, just for future reference I don't think its a good idea to continue using the method of modifying static display lists for dynamic behaviour - we should find a way to use dynamic DLs or pointers in static DLs.

kurethedead commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2711,6 +2752,133 @@ def save_binary(self, romfile, f3d, segments):
+            if triGroup.fMaterial:
+                bleed_mat = self.bleed_mat(triGroup.fMaterial, LastMat)
+                bleed_tex, bleed_scr = self.bleed_textures(triGroup.fMaterial, LastMat)
+            else:
+                bleed_mat = []
+                bleed_tex = []
```
By the way, you didn't actually fix this. I'm assuming you meant `bleed_scr = None`.

jesusyoshi54 commented:
Im not sure what the issue is with modifying Static data is as its something that's always been done for tile scrolls. It isn't like the DLs are literally labelled as Static in the output file.
If I were to use DL jumps or pointers in static DLs it would defeat the purpose if making an inline FMesh draw in the first place.

jesusyoshi54 commented:
Also more on why store the bled list, it is because I don't want to store set tile scroll indices to the inline DL, but a pointer to the material start. 

This also makes the code more flexible for other dynamic properties added in the future. If desired, there could easily be changes made to add dynamic palettes, or dynamic texture loads using the most of the same code, because the inlined material is easily accessible.

kurethedead commented:
Like I said, you can add a callback inside the triGroup loop in bleed() to do this. Whats the purpose of halfway calculating the index in bleed() and then saving intermediate display lists to finish calculating it again in tileScrollInlineMaterialToC() when you can just calculate it all at once in bleed()? Spreading index management through multiple locations just makes things confusing.

jesusyoshi54 commented:
If I remove the triGroup.bleed property I would have to do the full parsing of tile scroll indices inside the FMesh.bleed function.
I really do not like the idea of doing that because it goes against the established process for non bled gfx and does multiple operations all at once when it really should just be building the bled GfxLists.
If there was something like SM64_fMesh that used the callback to calculate tile scroll indices it would make sense, but otherwise I am adding sm64 specific code to the bleed operation to do this tile scroll, so I want to have that code in the GfxFormatter which requires the bled DL, and the location of the bled DL.

kurethedead commented:
Calculating indices is part of the process of building the GfxList. These are indices into the list that's being built. If you think tile scrolling is going to be unique to SM64 then you can add a callback in the inlining function that retrieves the correct indices, which gets overridden in SM64_Model.

jesusyoshi54 commented:
If it is part of the gfx building process why is it done after all c writing is done originally.
That reason I was trying to make it this way was to mirror the original process.

kurethedead commented:
Most likely because it was simple enough before that it didn't matter where it happened. I didn't write the tile scrolling so I don't know specifically why.

jesusyoshi54 commented:
Ok, so I was entirely mistaken about it.
What I can do is add in a mesh callback function that calculates these tile scroll indices that is called while the triGroup is being added to the fMesh.draw list.
Ill need to store these indices as a dict, and into the triGroup as a prop, replacing the triGroup.bleed attribute. Then the current inline tile scroll function will read these indices directly instead of parsing a DL.
Ultimately I still think it is better to store the indices to the triGroup over the mesh simply because the triGroup is already holding the info for the scroll with trGroup.fMaterial, so I avoid having a new mesh prop storing materials.
I'll also add a few more callback functions for overriding with subclasses.
Other comments on code besides the tile scrolling I will edit as suggested.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2807,8 +2975,15 @@ def get_f3d_mat_from_version(material: bpy.types.Material):
+        if inline:
+            self.textures = [GfxList(f"tex_{i}_" + name, GfxListTag.Material, DLFormat.Static) for i in range(2)]
+        else:
+            self.textures = [self.material] * 2
```
In this case, do you think self.textureDLs should be something like:
```python
        if inline:
            self.textureDLs = [GfxList(f"tex_{i}_" + name, GfxListTag.Material, DLFormat.Static) for i in range(2)]
        else:
            self.textureDLs = None
```
I feel like even if this intent is more clear, functionally it would be better to leave the mutability case up as the correct behavior would execute even when the getter function is not used.

KazeEmanuar commented:
before: https://cdn.discordapp.com/attachments/969282449609752577/1070868603517280346/2023-02-02_20-43-22.mp4

after: https://cdn.discordapp.com/attachments/969282449609752577/1070868619623403530/2023-02-02_20-48-05.mp4

it's a pretty good perf improvement and looks identical.

thecozies commented:
RCP crash for unknown reasons

https://user-images.githubusercontent.com/79979276/216644721-32c2b9c4-0e84-45a2-b528-bab737015405.mp4

___

Another issue, my scrolling textures a bit corrupted (also scrolling crazy fast imo)

![image](https://user-images.githubusercontent.com/79979276/216644820-c7764e3f-4d9d-4fa5-94f7-ec8f4a2d65fb.png)

https://user-images.githubusercontent.com/79979276/216645261-17be97a1-7bef-4145-98fc-545535e2888b.mp4

![image](https://user-images.githubusercontent.com/79979276/216645497-4ff88805-30ad-4601-993d-77a67ed73358.png)

___

heres the blend and model.inc.c (exports to BOB)

[blend_and_model_inc.zip](https://github.com/Fast-64/fast64/files/10580386/blend_and_model_inc.zip)

jesusyoshi54 commented:
uh oh.
Ill take a look, my guess is that this crash is scroll related.

thecozies commented:
Ahh another thing i noticed

![image](https://user-images.githubusercontent.com/79979276/216646558-40afdcd4-db15-4761-b728-d3879734c5d5.png)

looks like fog color and pos were set twice, there was a material in between those but it did not set fog color or position

thecozies commented:
WAIT I THINK IT WAS THE MAT STACK SCRIPT THAT BROKE IT

thecozies commented:
IT WAS

thecozies commented:
yup everything is fine so far... mat stack script being broken with it is a problem, but not a problem for this probably

https://user-images.githubusercontent.com/79979276/216649552-5d2cb58d-ca3b-4b1f-aec5-776a843afdbe.mp4

kurethedead reviewed (changes requested):
After reviewing this code I think that the bleed functionality should be moved out into its own class. The implementation is pretty hacky and creates a lot of cross cutting concerns. If we have a separate bleed object that takes in an FModel to process, then we can apply sm64 armature bleed after the geolayout is finalized (after overrides are generated). From what I've seen, it isn't necessary for bleed to be applied right after the completion of an FMesh, so it should be okay to do it after the completion of the entire FModel. All of my specific comments were made before the armature bleed was added.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2826,6 +2989,7 @@ def __init__(self, name, index, fMaterial):
         self.fMaterial = fMaterial
         self.vertexList = VtxList(name + "_vtx_" + str(index))
         self.triList = GfxList(name + "_tri_" + str(index), GfxListTag.Geometry, DLFormat.Static)
+        self.name = name
```
This was for when I was adding comments to the output to mark each material/tri list. It should be useless now though.

jesusyoshi54 replied:
This was for when I was adding comments to the output to mark each material/tri list. It should be useless now though.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -4463,29 +4686,34 @@ def gsDPLoadTileGeneric(c, tile, uls, ult, lrs, lrt):
+    # due to tile scrolls, bleed must be done by other checks
+    def bleed(self, lastGfxList: GfxList, curList: GfxList, curIndex: int):
+        return self._bleed
+
```
This has been fixed to be based on the cmd tags.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2809,13 +2833,152 @@ def save_binary(self, romfile, f3d, segments):
-        for triGroup in self.triangleGroups:
-            staticData.append(triGroup.to_c(f3d, gfxFormatter))
-        dynamicData = gfxFormatter.drawToC(f3d, self.draw)
+        if self.inline:
+            for triGroup in self.triangleGroups:
+                # add in vertex data
+                staticData.append(triGroup.vertexList.to_c())
+            staticData.append(gfxFormatter.drawToC(f3d, self.draw))
+            dynamicData = CData()
+        else:
+            for triGroup in self.triangleGroups:
+                staticData.append(triGroup.to_c(f3d, gfxFormatter))
+            dynamicData = gfxFormatter.drawToC(f3d, self.draw)
```
I don't think editing thd tri list itself is a good idea because the tri list is used multiple times when things are referenced several times (like in armatures).
I could add a BleedTag to the triList gfxList though.

kurethedead reviewed (commented):
I've made some comments about removing inline checks from FModel and its other contained objects. I can't comment on it since it wasn't a change in this PR, but for `FModel.to_c_gfx_scroll()`, we can change it to run the scroll processing on both `fMaterial.material` and `fMesh.draw` for all cases. To account for this, at the end of bleeding, `fMaterial.material` should be set to `None` (and all the generic export code should be modified to check for None). Along with the other change I suggest relating to `textureDLs`, there wouldn't even be an issue of losing access to material data by doing this.

kurethedead commented:
Sorry, I forgot to write out the main idea of the previous comment: Bleeding would set every FMaterial.material and FTriGroup.triList to None after completion. For SM64, this would be done after processing the geolayout, so there wouldn't be any issue with needing the triLists. If the FModel export process handles these None checks, then it doesn't need to keep track of whether it is inline or not.

jesusyoshi54 commented:
Why would I want to get rid of data as a way to keep track of my post processing instead of just using a bool or enum? All the inline checks are based on inheritance and I don't really think they're intrusive to the normal model creation process.

Also on a different note, while I don't plan on adding it this PR, there are requests and interest in making bleed a per mesh/bone feature, in which case I definitely would not want to delete any of the source cmd lists, and I would want to keep track of inline on this per fmesh/bone basis.

Though if having an inline property on the class hierarchy is not good, I could also simply attach a bleed tag onto the gfxList class, and then not write it in specific places based on that. A bleed tag would be something that the "bleeder" class would only write so it is not really adding extra information onto the class. I think I could also probably rework the bleed tags to simply be a a gfxList tag enum called "bleed" and use that instead. 
I'm not certain if I will not need the bleed tags as a way to keep track of bleed status (for revert purposes really) but I definitely think it is worth keeping the original data.

kurethedead commented:
You would get rid of the display lists because FModel is a representation of the final export, and in the final export that data doesn't exist. If you decide to make bleed per mesh/bone, then that could be a toggle on the geolayout node, not the mesh data itself. The source DLs would be deleted at the end of this process, so there isn't any need for them. As for the inline checks, I do think they are intrusive - people reading generalized export code have to also keep in mind the bleeding process while doing it even if they aren't intending to do anything with bleed. In my opinion it should be avoided if possible.

jesusyoshi54 commented:
Well if you want to get rid of the inline checks I still think it would be better to just change the gfx list tag rather than delete the gfx list itself.
I can make it something like, no export, or null.

kurethedead commented:
This is just pushing the complexity down to lower levels where it makes things more cluttered. In the future we could change it if we need it, but I'm not seeing in what case you would want to access these DLs after inlining them. The whole idea of inlining implies that it should be done once the higher level abstractions of materials/meshes aren't needed anymore.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -1749,6 +1749,9 @@ def tile_func(direction: str, speed: int, cmd_num: int):
 
+    if cmd_num == -1:
+        return None
+
```
Nope, resolved with gfx tag PR

jesusyoshi54 commented:
Ok I made the changes.
Now writing is determined by if the GfxList.commands is None or not, instead of if the inline bool exists.
I also took all the GbiMacro bleed funcs and put them in the BleedGraphics class.

Basically all the inline checks are removed throughout the repo.
fMaterial is not split up into mat_only_DL and texture_DLs, and then when the material is done being wrote to, they get combined into one under the fMaterial.material gfx list.

There are a few misc changes in here besides bleeding, such as annotations or changing some properties in GbiMacros to be more consistent (all tile props are called "tile" instead of some being "tile" and others being "t")
I also applied black to f3d_gbi.py

kurethedead commented on fast64_internal/sm64/sm64_geolayout_classes.py:
```diff
@@ -447,6 +448,35 @@ def to_c(self):
 
+class GeoLayoutBleed(BleedGraphics):
+    def bleed_geo_layout_graph(self, fModel, geo_layout_graph):
+        if not fModel.inline:
+            return
```
This happens multiple times, but with `inline` removed these checks should be moved outside these classes.

kurethedead reviewed (changes requested):
The latest changes are an improvement. I made a couple more comments about decoupling bleeding from the f3d_gbi classes. I also think that at this point all the bleed code in `f3d_gbi.py` should be moved into its own file.

jesusyoshi54 commented:
You commented it in several places so Ill just make a general reply.
The mat write method is only set to WriteDifferentAndRevert explicitly for sm64. For OOT it is explicitly set to WriteAll, and for generic f3d exports it is based on your drop down selection of a scene prop.
With this in mind I believe using that in lieu of an inline bool is not proper, as we shouldn't force users into using this write method based on they choose to have materials are written.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2821,6 +2840,289 @@ def to_c(self, f3d, gfxFormatter):
+        self.clear_gfx_lists(fModel)
+
+    # clear the gfx lists so they don't export
+    def clear_gfx_lists(self, fModel):
+        for (fMaterial, texDimensions) in fModel.materials.values():
+            fMaterial.material.commands = None
+            fMaterial.revert = None
+        for fMesh in fModel.meshes.values():
+            for tri_list in fMesh.triangleGroups:
+                tri_list.triList.commands = None
```
Yeah it is about the same amount of work but I was weighing options.
If I clear the gfx list, you will get exceptions in any place you don't do a None check, but it might be clearer in intention. So adding may this requires more checks in various places and possible issues if specific subclasses try to use their own write methods.
Clearing the command list makes it easy to check inside the GfxList as a central location, and maybe even carries over into cases where you don't clear the list but it has no commands anyway.
If it is better to do the gfx list itself I could easily do that though.

kurethedead replied:
I think its better to push the complexity upwards into the owning classes (`FMaterial` and `FTriGroup`). Even if you have to handle more specific cases, this lets you use `GfxList` for other purposes without having to worry about null checking on the commands.

jesusyoshi54 commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2821,6 +2840,289 @@ def to_c(self, f3d, gfxFormatter):
+        return True
+
+    def bleed_SPSetOtherMode(self, cmd_list: GfxList, cmd: GbiMacro):
+        if self.bleed_state != self.BleedTag.BleedStart:
```
Ok I have an idea that may be more clear.
I could use the BleedGfxLists class to store this info, and then pass that class to each callback. That class will hold the info for that specific bled GfxList so I can store any props I need about the bleed state there. Then I can rename the bool to something specific to this interaction, like `bleed_other_mode`

kurethedead replied:
I'm not sure why you consider the argument "unused", the only case that would be true is if there's no other mode command. The names are clear enough that their general intent can be understood, but the point is that if a maintainer wants to make code changes and avoid/fix buggy behaviour, they have to know the exact points in code where the changes happen, which is a unhelpful if the person doesn't know the bleed code inside and out. Ultimately the location you are in the code should be enough knowledge about the current bleed state, and having a separate variable like this makes the code more difficult to work around. I'm not sure about the other solution you're proposing either, it seems like a more complicated version of what I was proposing.

jesusyoshi54 commented:
Ok I changed the construction of the FModel classes so that the passed mat write method is what gets used.
I also moved the bleed class to its own file, and made some of the changes recommended to the bleed class itself.
I also did some slight renaming of variables inside the bleed class, but ultimately nothing has functionally changed.

kurethedead reviewed (commented):
The `inline` variables/checks should be removed from FModel and any subclasses of it. The `inline` checks happening at the beginning of the bleed functions can be moved right outside of those functions. `SM64Model` can have the write method passed to it instead of `inline`.

jesusyoshi54 commented:
Ok I removed all uses of inline as a property and made it a bool either passed to a function, or used as a ternary for mat write methods.
SM64Model now also has matWriteMethod passed to it as a constructing argument.

kurethedead reviewed (commented):
The gfx scroll inline arguments can be removed as well.

kurethedead commented on fast64_internal/f3d/f3d_gbi.py:
```diff
@@ -2580,9 +2594,9 @@ def to_c_vertex_scroll(self, gfxFormatter: GfxFormatter) -> CScrollData:
-    def to_c_gfx_scroll(self, gfxFormatter: GfxFormatter) -> CScrollData:
+    def to_c_gfx_scroll(self, gfxFormatter: GfxFormatter, inline: bool = False) -> CScrollData:
         data = CScrollData()
-        if not self.inline:
+        if not inline:
             for _, (fMaterial, _) in self.materials.items():
                 fMaterial: FMaterial
                 data.append(gfxFormatter.gfxScrollToC(fMaterial.material, self.f3d))
```
Both `fMaterial.material` and `fMesh.draw` can be handled (works for either bleed on/off) as long as `fMaterial.material` is None checked, letting you remove `inline` from this.

jesusyoshi54 replied:
good call

kurethedead reviewed (commented):
Whoops, forgot my attached comment.

kurethedead commented:
I made a PR to this branch to fix a minor naming duplicate issue with scrolling variables. However, I'm still getting an issue where textures aren't showing up with bleed enabled. This seems to happen if exporting with microcode version F3D, regardless if the repo itself is F3D or F3DEX2. When exporting F3DEX2 the issue is gone.
<img width="622" alt="WpPMrWLYXO" src="https://user-images.githubusercontent.com/59840896/219821746-bf4b7d50-e906-4174-8246-f48fe5fa3f54.png">
<img width="602" alt="blender_57OiYTEf7o" src="https://user-images.githubusercontent.com/59840896/219821752-2ee37da9-db52-42ed-ac96-7d43e4b3d91b.png">

jesusyoshi54 commented:
I can't replicate this issue. Could you post the exported mesh DL during the F3D export?

kurethedead commented:
```
Gfx bob_dl_Plane_001_mesh_layer_1[] = {
	gsDPPipeSync(),
	gsDPSetCombineLERP(TEXEL0, 0, SHADE, 0, 0, 0, 0, ENVIRONMENT, TEXEL0, 0, SHADE, 0, 0, 0, 0, ENVIRONMENT),
	gsSPSetOtherMode(G_SETOTHERMODE_H, 4, 20, G_AD_NOISE | G_CD_MAGICSQ | G_CK_NONE | G_TC_FILT | G_TF_BILERP | G_TL_TILE | G_TD_CLAMP | G_TP_PERSP | G_CYC_1CYCLE | G_PM_1PRIMITIVE),
	gsSPSetOtherMode(G_SETOTHERMODE_L, 0, 3, G_AC_NONE | G_ZS_PIXEL),
	gsSPSetLights1(bob_dl_f3dlite_material_005_lights),
	gsDPSetTextureLUT(G_TT_RGBA16),
	gsDPSetTextureImage(G_IM_FMT_RGBA, G_IM_SIZ_16b, 1, bob_dl_eyes_closed_ci8_ci8_pal_rgba16),
	gsDPSetTile(0, 0, 0, 256, 7, 0, G_TX_WRAP | G_TX_NOMIRROR, 0, 0, G_TX_WRAP | G_TX_NOMIRROR, 0, 0),
	gsDPLoadTLUTCmd(7, 91),
	gsDPLoadSync(),
	gsDPSetTextureImage(G_IM_FMT_CI, G_IM_SIZ_8b_LOAD_BLOCK, 1, bob_dl_eyes_closed_ci8_ci8),
	gsDPSetTile(G_IM_FMT_CI, G_IM_SIZ_8b_LOAD_BLOCK, 0, 0, 7, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0, G_TX_WRAP | G_TX_NOMIRROR, 6, 0),
	gsDPLoadBlock(7, 0, 0, 1023, 256),
	gsDPSetTile(G_IM_FMT_CI, G_IM_SIZ_8b, 8, 0, 0, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0, G_TX_WRAP | G_TX_NOMIRROR, 6, 0),
	gsDPSetTileSize(0, 0, 0, 252, 124),
	gsSPVertex(bob_dl_Plane_001_mesh_layer_1_vtx_0 + 0, 4, 0),
	gsSP1Triangle(0, 1, 2, 0),
	gsSP1Triangle(0, 2, 3, 0),
	gsDPPipeSync(),
	gsDPSetCombineLERP(TEXEL0, 0, TEXEL1, 0, 0, 0, 0, 1, COMBINED, 0, SHADE, 0, 0, 0, 0, 1),
	gsSPSetOtherMode(G_SETOTHERMODE_H, 4, 20, G_AD_NOISE | G_CD_MAGICSQ | G_CK_NONE | G_TC_FILT | G_TF_BILERP | G_TL_TILE | G_TD_CLAMP | G_TP_PERSP | G_CYC_2CYCLE | G_PM_1PRIMITIVE),
	gsSPSetOtherMode(G_SETOTHERMODE_L, 0, 32, G_AC_NONE | G_ZS_PIXEL | G_RM_PASS | G_RM_AA_ZB_OPA_SURF2),
	gsSPSetLights1(bob_dl_f3dlite_material_001_layer1_lights),
	gsDPSetTextureLUT(G_TT_NONE),
	gsDPSetTextureImage(G_IM_FMT_RGBA, G_IM_SIZ_16b_LOAD_BLOCK, 1, bob_dl_b_rgba16),
	gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b_LOAD_BLOCK, 0, 0, 7, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0),
	gsDPLoadBlock(7, 0, 0, 1023, 256),
	gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b, 8, 0, 0, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0),
	gsDPSetTileSize(0, 0, 0, 124, 124),
	gsDPSetTextureLUT(G_TT_NONE),
	gsDPSetTextureImage(G_IM_FMT_RGBA, G_IM_SIZ_16b_LOAD_BLOCK, 1, bob_dl_a_rgba16),
	gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b_LOAD_BLOCK, 0, 256, 6, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0),
	gsDPLoadBlock(6, 0, 0, 1023, 256),
	gsDPSetTile(G_IM_FMT_RGBA, G_IM_SIZ_16b, 8, 256, 1, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0, G_TX_WRAP | G_TX_NOMIRROR, 5, 0),
	gsDPSetTileSize(1, 0, 0, 124, 124),
	gsSPVertex(bob_dl_Plane_001_mesh_layer_1_vtx_1 + 0, 4, 0),
	gsSP1Triangle(0, 1, 2, 0),
	gsSP1Triangle(0, 2, 3, 0),
	gsDPSetCycleType(G_CYC_1CYCLE),
	gsDPSetTextureLUT(G_TT_NONE),
	gsDPSetRenderMode(G_RM_AA_ZB_OPA_SURF, G_RM_AA_ZB_OPA_SURF2),
	gsDPPipeSync(),
	gsSPEndDisplayList(),
};
```

kurethedead commented:
```
const GeoLayout bob_area_1_geo[] = {
	GEO_NODE_START(),
	GEO_OPEN_NODE(),
		GEO_SWITCH_CASE(5, geo_switch_area),
		GEO_OPEN_NODE(),
			GEO_BRANCH(1, bob_dl_Room1_geo),
			GEO_BRANCH(1, bob_dl_Room2_geo),
			GEO_BRANCH(1, bob_dl_Room3_geo),
			GEO_BRANCH(1, bob_dl_Room4_geo),
		GEO_CLOSE_NODE(),
		GEO_TRANSLATE_NODE(LAYER_OPAQUE, 10, -306, 425),
	GEO_CLOSE_NODE(),
	GEO_RETURN(),
};
const GeoLayout bob_dl_Room1_geo[] = {
	GEO_NODE_START(),
	GEO_OPEN_NODE(),
		GEO_TRANSLATE_NODE(LAYER_OPAQUE, 0, -300, 0),
		GEO_OPEN_NODE(),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, 0, -714, 0, bob_dl_Plane_mesh_layer_1),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, 51, -433, 617, bob_dl_Plane_004_mesh_layer_1),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, 896, -714, 180, bob_dl_Plane_008_mesh_layer_1),
		GEO_CLOSE_NODE(),
	GEO_CLOSE_NODE(),
	GEO_RETURN(),
};
const GeoLayout bob_dl_Room2_geo[] = {
	GEO_NODE_START(),
	GEO_OPEN_NODE(),
		GEO_TRANSLATE_NODE(LAYER_OPAQUE, 0, -478, -1323),
		GEO_OPEN_NODE(),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, 0, -714, 0, bob_dl_Plane_001_mesh_layer_1),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, -1055, -714, 118, bob_dl_Plane_006_mesh_layer_1),
		GEO_CLOSE_NODE(),
	GEO_CLOSE_NODE(),
	GEO_RETURN(),
};
const GeoLayout bob_dl_Room3_geo[] = {
	GEO_NODE_START(),
	GEO_OPEN_NODE(),
		GEO_TRANSLATE_NODE(LAYER_OPAQUE, 0, -712, -2760),
		GEO_OPEN_NODE(),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, 0, -714, 0, bob_dl_Plane_002_mesh_layer_1),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, 1110, -714, 142, bob_dl_Plane_005_mesh_layer_1),
		GEO_CLOSE_NODE(),
	GEO_CLOSE_NODE(),
	GEO_RETURN(),
};
const GeoLayout bob_dl_Room4_geo[] = {
	GEO_NODE_START(),
	GEO_OPEN_NODE(),
		GEO_TRANSLATE_NODE(LAYER_OPAQUE, 0, -1057, -4202),
		GEO_OPEN_NODE(),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, 0, -714, 0, bob_dl_Plane_003_mesh_layer_1),
			GEO_TRANSLATE_NODE_WITH_DL(LAYER_OPAQUE, -791, -714, -32, bob_dl_Plane_007_mesh_layer_1),
		GEO_CLOSE_NODE(),
	GEO_CLOSE_NODE(),
	GEO_RETURN(),
};
const GeoLayout bob_area_1[] = {
	GEO_NODE_SCREEN_AREA(10, SCREEN_WIDTH/2, SCREEN_HEIGHT/2, SCREEN_WIDTH/2, SCREEN_HEIGHT/2),
	GEO_OPEN_NODE(),
		GEO_ZBUFFER(0),
		GEO_OPEN_NODE(),
			GEO_NODE_ORTHO(100.0000),
			GEO_OPEN_NODE(),
				GEO_BACKGROUND(BACKGROUND_OCEAN_SKY, geo_skybox_main),
			GEO_CLOSE_NODE(),
		GEO_CLOSE_NODE(),
		GEO_ZBUFFER(1),
		GEO_OPEN_NODE(),
			GEO_CAMERA_FRUSTUM_WITH_FUNC(45.0000, 100, 30000, geo_camera_fov),
			GEO_OPEN_NODE(),
				GEO_CAMERA(CAMERA_MODE_8_DIRECTIONS, 0, -348, 0, 0, -348, 0, geo_camera_main),
				GEO_OPEN_NODE(),
					GEO_BRANCH(1, bob_area_1_geo),
					GEO_RENDER_OBJ(),
					GEO_ASM(ENVFX_MODE_NONE, geo_envfx_main),
				GEO_CLOSE_NODE(),
			GEO_CLOSE_NODE(),
		GEO_CLOSE_NODE(),
		GEO_DISPLAY_LIST(LAYER_OPAQUE, bob_dl_material_revert_render_settings),
	GEO_CLOSE_NODE(),
	GEO_END(),
};
```

jesusyoshi54 commented:
The issue was related to the usage of rooms. I added a boolean to the geo layout bleed function to change how last used materials are bleed in the case where rooms are used.

kurethedead reviewed (approved):
Seems good to me, I'll merge this soon unless there are any other issues that come up.
